Better sign/verify tests, eddsa support, refactoring of the KeyProvider api

gasbytes opened the pull request:
This PR includes the following changes:
- support for e25519 for signing and verifying;
- added sign/verify tests in the e2e test file (ecdsa, eddsa, rsa) for more coverage this allowed to discover a couple of bugs during the loading of the private key used for signing;
- generic signer for rsapss;
- support for rsa-pkcs1-sha512;
- removed pub keywords on various kx structs since they are not needed;
- various bug fixes;

gasbytes committed:
Initial support for ed25519, working on ed448, and then will work on the unit
test.

gasbytes committed:
ed448 support, working on the tests

gasbytes committed:
working on the eddsa tests (also refactoring) staging to work on the linux machine.

gasbytes committed:
checking stat instead of ret when verifying

gasbytes committed:
correctly falling back to the right signing/verifying scheme in case of error
(try_from logic).

gasbytes committed:
Signing and verifying ecdsa (all curves) e2e test completed & fixed some bugs
in their respective implementation.

gasbytes committed:
eddsa tests

gasbytes committed:
- Generic Signer for RsaPSS
- completed the signing/verifying e2e test for RsaPSS
- fixed bugs during the process
- now replicating for RsaPKCS1

gasbytes committed:
sign/verify rsa-pkcs1

gasbytes committed:
support rsa-pkcs1-sha512

gasbytes committed:
removed 'pub' keywords from kx structs

gasbytes committed:
Minor changes
- refactor of ed25519 signer
- minor changes (using macros instead of fixed values in various files)

JacobBarthelmeh commented on rustls-wolfcrypt-provider/src/sign/eddsa.rs:
```diff
@@ -0,0 +1,169 @@
+    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, rustls::Error> {
+        let mut ret;
+        let message_length: word32 = message.len() as word32;
+        let mut sig: [u8; 1024] = [0; 1024];
```
avoid magic numbers, why 1024? There is a ED25519_SIG_SIZE macro for the size needed.

JacobBarthelmeh commented on rustls-wolfcrypt-provider/src/verify/rsapss.rs:
```diff
@@ -152,39 +176,86 @@ impl SignatureVerificationAlgorithm for RsaPssSha384Verify {
+        signature: &[u8],
+    ) -> Result<(), InvalidSignature> {
+        let mut digest: [u8; WC_SHA512_DIGEST_SIZE as usize] = [0; WC_SHA512_DIGEST_SIZE as usize];
+        let mut out: [u8; 256] = [0; 256];
```
Will this fail if used with 4096 bit RSA keys? Seems hard set to only handle 2048 bit signatures. If so can we make an adjustment here to handle the max RSA key size supported.

JacobBarthelmeh reviewed (changes requested):
Couple minor items, tests ran and passed as expected. For future pull requests, please open it from a fork of the repository rather than a branch off of the wolfSSL repository.

gasbytes committed:
Removed magic number for maximum EDDSA signature size, using ED25519_SIG_SIZE instead.

gasbytes committed:
Addressing review:
- Updated the sizes of the signature buffer to support up to 4096 bits
  rsa keys;
- Updated to e2e.rs test to test 2048 and 4096 bits rsa keys;

gasbytes committed:
tests: Add parallel execution and mutex-based server access

- Parallelize RSA PSS and PKCS1 testing with rayon, reducing PSS time
 by 52% and PKCS1 by 75%
- Replace #[serial] with mutex-based server access, gain a some
  seconds on the total test time.

gasbytes committed:
Merge branch 'main' into better-sign-verify-tests

gasbytes committed:
Linux-specific fixes
- Fixed SEGV on linux build (stack misalignment issue)
- Fixed minor rayon initialization bug
- Updated .gitignore and added missing package (hex_literal)
- platform specific changes for rsa_pss_sign_and_verify test
- Removed build.sh (conflict resolution)
